KAFKA-19221: Propagate IOException on LogSegment#close

gaurav-narula opened the pull request:
Log segment closure results in right sizing the segment on disk along
with the associated index files.

This is specially important for TimeIndexes where a failure to right
size may eventually cause log roll failures leading to under replication
and log cleaner failures.

This change uses `Utils.closeAll` which propagates exceptions, resulting
in an "unclean" shutdown. That would then cause the broker to attempt to
recover the log segment and the index on next startup, thereby avoiding
the failures described above.

Reviewers: Omnia Ibrahim <o.g.h.ibrahim@gmail.com>, Jun Rao
 <junrao@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

gaurav-narula committed:
KAFKA-19221: Propagate IOException on LogSegment#close

Log segment closure results in right sizing the segment on disk along
with the associated index files.

This is specially important for TimeIndexes where a failure to right
size may eventually cause log roll failures leading to under replication
and log cleaner failures.

This change uses `Utils.closeAll` which propagates exceptions, resulting
in an "unclean" shutdown. That would then cause the broker to attempt to
recover the log segment and the index on next startup, thereby avoiding
the failures described above.

gaurav-narula commented:
CC: @soarez @OmniaGM

gaurav-narula commented:
CC: @chia7712 can you please take a look?

OmniaGM reviewed (approved):
Nice improvement. This looks like a small change. @ijuma & @chia7712 can one of you please review this change?

junrao commented on server/src/test/java/org/apache/kafka/server/LogManagerIntegrationTest.java:
```diff
@@ -57,6 +58,70 @@ public LogManagerIntegrationTest(ClusterInstance cluster) {
         this.cluster = cluster;
     }
 
+    @ClusterTest(types = {Type.KRAFT}, brokers = 3)
```
Could we just use 1 broker in the test?

junrao commented on server/src/test/java/org/apache/kafka/server/LogManagerIntegrationTest.java:
```diff
@@ -57,6 +58,70 @@ public LogManagerIntegrationTest(ClusterInstance cluster) {
+        assertTrue(timeIndexFile.exists());
+        assertTrue(timeIndexFile.setReadOnly());
+
+        cluster.brokers().get(0).shutdown();
```
Could we verify that the cleanShutdown file is not written?

junrao reviewed (commented):
@gaurav-narula : Thanks for the PR. Left a couple of comments.

gaurav-narula committed:
Address review comments

gaurav-narula commented:
Thanks for the review @junrao. I've addressed your comments with [2677f90](https://github.com/apache/kafka/pull/19607/commits/2677f90c44cd675aac9270eb1f7a36497ebdd88c) and rebased against trunk.

chia7712 commented on server/src/test/java/org/apache/kafka/server/LogManagerIntegrationTest.java:
```diff
@@ -57,6 +59,68 @@ public LogManagerIntegrationTest(ClusterInstance cluster) {
+
+        cluster.brokers().get(0).shutdown();
+
+        assertEquals(1, cluster.brokers().get(0).config().logDirs().size());
```
Could you please add a variable for `cluster.brokers().get(0)`?
```scala
        var broker = cluster.brokers().get(0);
        broker.shutdown();

        assertEquals(1, broker.config().logDirs().size());
        String logDir = broker.config().logDirs().get(0);
```

chia7712 commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java:
```diff
@@ -751,10 +751,7 @@ public Optional<FileRecords.TimestampAndOffset> findOffsetByTimestamp(long times
-        Utils.closeQuietly(lazyTimeIndex, "timeIndex", LOGGER);
-        Utils.closeQuietly(log, "log", LOGGER);
-        Utils.closeQuietly(txnIndex, "txnIndex", LOGGER);
+        Utils.closeAll(lazyOffsetIndex, lazyTimeIndex, log, txnIndex);
```
If `LogSegment#close` now throws an exception, then `LogSegments#close` might break without closing all segments, right?

chia7712 replied:
https://github.com/apache/kafka/blob/4eac6adf91a1516f8862dc5b40ca1dfe46135695/storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegments.java#L108

gaurav-narula replied:
Thanks for pointing that! I've modified `LogSegments#close` to ensure we close all log segments even if one of them throws and added a test in `LogSegmentsTest`.

In the process, I found that some tests assumed that resources like Indexes and FileRecords may be closed multiple times. With `LogSegment#close` now propagating exceptions, we need to exit early if the resources have been closed before, otherwise we'd see failures due to FileChannel being closed or mmap being `null`. I've therefore updated `AbstractIndex`, `TransactionIndex` and `FileRecords` appropriately.

chia7712 reviewed (commented):
@gaurav-narula thanks for this fix.

gaurav-narula committed:
Extract broker variable and ensure we close all log segments

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogSegmentsTest.java:
```diff
@@ -274,4 +278,27 @@ public void testUpdateDir() throws IOException {
+
+        doThrow(new IOException("Failure")).when(seg2).close();
+
+        try {
```
Could you please consider using `assertThrows`?

gaurav-narula replied:
Addressed with [ce22ba9](https://github.com/apache/kafka/pull/19607/commits/ce22ba92e1ab3c9da057eef6e997ad6c3534e5f9)

chia7712 reviewed (commented):
@gaurav-narula thanks for updates. one small comment is left. PTAL

gaurav-narula committed:
Use assertThrows

chia7712 commented:
```
* What went wrong:
Execution failed for task ':storage:spotlessJavaCheck'.
> The following files had format violations:
      src/test/java/org/apache/kafka/storage/internals/log/LogSegmentsTest.java
          @@ -20,6 +20,7 @@
           import·org.apache.kafka.common.utils.Time;
           import·org.apache.kafka.common.utils.Utils;
           import·org.apache.kafka.test.TestUtils;
          +
           import·org.junit.jupiter.api.AfterEach;
           import·org.junit.jupiter.api.BeforeEach;
           import·org.junit.jupiter.api.Test;
  Run './gradlew :storage:spotlessApply' to fix these violations.

 ```
@gaurav-narula could you please fix the build error?

gaurav-narula committed:
Spotless apply

gaurav-narula commented:
@chia7712 should be good now

junrao reviewed (approved):
@gaurav-narula : Thanks for the updated PR. LGTM

chia7712 commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java:
```diff
@@ -751,10 +751,7 @@ public Optional<FileRecords.TimestampAndOffset> findOffsetByTimestamp(long times
     public void close() throws IOException {
         if (maxTimestampAndOffsetSoFar != TimestampOffset.UNKNOWN)
             Utils.swallow(LOGGER, Level.WARN, "maybeAppend", () -> timeIndex().maybeAppend(maxTimestampSoFar(), shallowOffsetOfMaxTimestampSoFar(), true));
```
I think writing the max timestamp has a similar issue. the last entry is assumed to be the max timestamp after restarting. Hence, it needs to rebuild the index if the "true" max timestamp is not stored correctly

I open KAFKA-19428 to trace it
